File: src/voxam/glulx/gestalt.py
"""Gestalt selectors: what this interpreter can do (Glulx: Gestalt).

The reference glulxe answers most of these from compile-time
switches. Voxam answers them from Capabilities, a runtime value --
partly because Python has no compile step, and partly because it
lets the capability set track which eras exist yet. Every False in
the defaults is not a design decision but a statement that the
supporting era has not arrived, with the era named beside it; the
branch that builds an era flips its flag.
"""

from dataclasses import dataclass
from enum import IntEnum
from importlib.metadata import version
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from voxam.glulx.machine import Machine

# The Glulx specification version implemented: 3.1.3, packed as the
# header packs it (Glulx: The Header).
GLULX_VERSION = 0x00030103

MAJOR_SHIFT = 16
MINOR_SHIFT = 8


class Gestalt(IntEnum):
    """The selector numbers the gestalt opcode answers."""

    GLULX_VERSION = 0
    TERP_VERSION = 1
    RESIZE_MEM = 2
    UNDO = 3
    IO_SYSTEM = 4
    UNICODE = 5
    MEM_COPY = 6
    MALLOC = 7
    MALLOC_HEAP = 8
    ACCELERATION = 9
    ACCEL_FUNC = 10
    FLOAT = 11
    EXT_UNDO = 12
    DOUBLE = 13


# The io systems the IO_SYSTEM selector is asked about.
IOSYS_NULL = 0
IOSYS_FILTER = 1
IOSYS_GLK = 2


@dataclass(frozen=True)
class Capabilities:
    """What this build of the machine can currently do.

    Attributes:
        resize_mem: setmemsize works; the memory era built it.
        mem_copy: mzero and mcopy work; the exec-loop era.
        unicode: E2 strings, the wide nodes, streamunichar, and
            the type-14 stubs; the strings era.
        undo: saveundo and restoreundo; the save era carried them.
        ext_undo: hasundo and discardundo; the same era.
        malloc: malloc and mfree; the heap era carried them.
        acceleration: The accel era.
        floats: The float era.
        doubles: The double era.
        glk: A Glk library is installed on this machine; the
            bridge answers the glk opcode and iosys mode 2 works.
    """

    resize_mem: bool = True
    mem_copy: bool = True
    unicode: bool = True
    undo: bool = True
    ext_undo: bool = True
    malloc: bool = True
    acceleration: bool = False
    floats: bool = False
    doubles: bool = False
    glk: bool = False


def terp_version() -> int:
    """Voxam's own version, packed the way the header packs one.

    Read from the installed package so the answer can never drift
    from pyproject: release 1.2.3 answers 0x00010203.
    """

    major, minor, patch = (int(part) for part in version("voxam").split("."))

    return (major << MAJOR_SHIFT) | (minor << MINOR_SHIFT) | patch


def answer(  # noqa: PLR0911, PLR0912 -- one flat return per selector
    machine: "Machine", selector: int, argument: int
) -> int:
    """One gestalt query, answered honestly.

    Unknown selectors answer zero rather than erring: that is how
    a program written against a future spec probes an older
    interpreter (Glulx: Gestalt).
    """

    caps = machine.capabilities

    if selector == Gestalt.GLULX_VERSION:
        return GLULX_VERSION

    if selector == Gestalt.TERP_VERSION:
        return terp_version()

    if selector == Gestalt.RESIZE_MEM:
        return int(caps.resize_mem)

    if selector == Gestalt.UNDO:
        return int(caps.undo)

    if selector == Gestalt.IO_SYSTEM:
        # The null and filter systems always work; Glk is its own
        # era's promise to keep.
        if argument in (IOSYS_NULL, IOSYS_FILTER):
            return 1

        if argument == IOSYS_GLK:
            return int(caps.glk)

        return 0

    if selector == Gestalt.UNICODE:
        return int(caps.unicode)

    if selector == Gestalt.MEM_COPY:
        return int(caps.mem_copy)

    if selector == Gestalt.MALLOC:
        return int(caps.malloc)

    if selector == Gestalt.MALLOC_HEAP:
        # The heap's start address, or zero with no blocks extant
        # (Glulx: Gestalt).
        return machine.heap.start

    if selector == Gestalt.ACCELERATION:
        return int(caps.acceleration)

    if selector == Gestalt.ACCEL_FUNC:
        # No accelerated function is available until that era.
        return 0

    if selector == Gestalt.FLOAT:
        return int(caps.floats)

    if selector == Gestalt.EXT_UNDO:
        return int(caps.ext_undo)

    if selector == Gestalt.DOUBLE:
        return int(caps.doubles)

    return 0

File: csharp/Voxam.Core/Glulx/Gestalt.cs
using System.Reflection;

namespace Voxam.Core.Glulx;

/// <summary>The selector numbers the gestalt opcode answers (Glulx: Gestalt).</summary>
public enum Selector : uint
{
    GlulxVersion = 0,
    TerpVersion = 1,
    ResizeMem = 2,
    Undo = 3,
    IoSystem = 4,
    Unicode = 5,
    MemCopy = 6,
    Malloc = 7,
    MallocHeap = 8,
    Acceleration = 9,
    AccelFunc = 10,
    Floats = 11,
    ExtUndo = 12,
    Doubles = 13,
}

/// <summary>
/// What this build of the machine can currently do.
///
/// Every false below is not a design decision but a statement that
/// the supporting era has not arrived, with the era named beside it;
/// the branch that builds an era flips its flag.
/// </summary>
public sealed record Capabilities
{
    /// <summary>setmemsize works; the memory era built it.</summary>
    public bool ResizeMem { get; init; } = true;

    /// <summary>mzero and mcopy work; the execution-loop era.</summary>
    public bool MemCopy { get; init; } = true;

    /// <summary>E2 strings, the wide nodes and streamunichar; the strings era.</summary>
    public bool Unicode { get; init; } = true;

    /// <summary>malloc and mfree; the heap era carried them.</summary>
    public bool Malloc { get; init; } = true;

    /// <summary>accelfunc and accelparam; the acceleration era carried them.</summary>
    public bool Acceleration { get; init; } = true;

    /// <summary>saveundo and restoreundo; the save era carried them.</summary>
    public bool Undo { get; init; } = true;

    /// <summary>hasundo and discardundo; the same era carried them.</summary>
    public bool ExtUndo { get; init; } = true;

    /// <summary>The single-precision opcodes; the float era carried them.</summary>
    public bool Floats { get; init; } = true;

    /// <summary>The double-precision opcodes; the same era carried them.</summary>
    public bool Doubles { get; init; } = true;

    /// <summary>A Glk library is installed, so the Glk output system works.</summary>
    public bool Glk { get; init; }
}

/// <summary>
/// Gestalt selectors: what this interpreter can do (Glulx: Gestalt).
///
/// The reference glulxe answers most of these from compile-time
/// switches; these are answered from a runtime value instead, which
/// lets the capability set track which eras exist yet.
/// </summary>
public static class Gestalt
{
    /// <summary>
    /// The Glulx specification version implemented: 3.1.3, packed as
    /// the header packs it (Glulx: The Header).
    /// </summary>
    public const uint GlulxVersion = 0x00030103;

    private const int MajorShift = 16;
    private const int MinorShift = 8;

    // The io systems the IoSystem selector is asked about.
    private const uint IoSysNull = 0;
    private const uint IoSysFilter = 1;
    private const uint IoSysGlk = 2;

    /// <summary>
    /// This interpreter's own version, packed the way the header packs
    /// one: release 1.2.3 answers 0x00010203. Read off the assembly,
    /// so the answer can never drift from what the port versions at.
    /// </summary>
    public static uint TerpVersion
    {
        get
        {
            // The build stamps a version on every assembly, so there
            // is always one here to read.
            var version = typeof(Gestalt).Assembly.GetName().Version!;

            return ((uint)version.Major << MajorShift) | ((uint)version.Minor << MinorShift) | (uint)version.Build;
        }
    }

    /// <summary>
    /// One gestalt query, answered honestly. Unknown selectors answer
    /// zero rather than erring: that is how a program written against
    /// a future specification probes an older interpreter (Glulx:
    /// Gestalt).
    /// </summary>
    public static uint Answer(Machine machine, uint selector, uint argument) => (Selector)selector switch
    {
        Selector.GlulxVersion => GlulxVersion,
        Selector.TerpVersion => TerpVersion,
        Selector.ResizeMem => Flag(machine.Capabilities.ResizeMem),
        Selector.Undo => Flag(machine.Capabilities.Undo),
        Selector.IoSystem => IoSystem(machine, argument),
        Selector.Unicode => Flag(machine.Capabilities.Unicode),
        Selector.MemCopy => Flag(machine.Capabilities.MemCopy),
        Selector.Malloc => Flag(machine.Capabilities.Malloc),
        // The heap's start address, or zero with no blocks extant.
        Selector.MallocHeap => (uint)machine.Heap.Start,
        Selector.Acceleration => Flag(machine.Capabilities.Acceleration),
        // Per function: which numbers this interpreter can replace.
        Selector.AccelFunc => Flag(Accelerator.Available.Contains(argument)),
        Selector.Floats => Flag(machine.Capabilities.Floats),
        Selector.ExtUndo => Flag(machine.Capabilities.ExtUndo),
        Selector.Doubles => Flag(machine.Capabilities.Doubles),
        _ => 0,
    };

    // The null and filter systems always work; Glk is its own era's
    // promise to keep.
    private static uint IoSystem(Machine machine, uint argument) => argument switch
    {
        IoSysNull or IoSysFilter => 1,
        IoSysGlk => Flag(machine.Capabilities.Glk),
        _ => 0,
    };

    private static uint Flag(bool answer) => answer ? 1u : 0u;
}
